app/src/org/commcare/android/tasks/FormRecordCleanupTask.java

```diff
@@ -263,38 +234,31 @@ public void commit(String[] meta) throws IOException, SessionUnavailableExceptio
             throw new RuntimeException("Invalid Key Data while attempting to decode form submission for processing");
         }
 
-        //Construct parser for this form's internal data.
+        // Construct parser for this form's internal data.
         DataModelPullParser parser = new DataModelPullParser(is, factory);
+
+        // populate uuid, modified, and caseIDs arrays by parsing
         parser.parse();
 
-        //TODO: We should be committing all changes to form record models via the ASW objects, not manually.
-        FormRecord parsed = new FormRecord(r.getInstanceURI().toString(), newStatus, r.getFormNamespace(), r.getAesKey(),uuid[0], modified[0]);
+        fis.close();
+        is.close();
+
+        // TODO: We should be committing all changes to form record models via
+        // the ASW objects, not manually.
+        FormRecord parsed = new FormRecord(r.getInstanceURI().toString(),
+                newStatus, r.getFormNamespace(), r.getAesKey(),
+                uuid[0], modified[0]);
         parsed.setID(r.getID());
-        
-        //TODO: The platform adds a lot of unfortunate coupling here. Should split out the need to parse completely 
-        //uninitialized form records somewhere else.
-        
-        if(caseIDs[0] != null && r.getStatus().equals(FormRecord.STATUS_UNINDEXED)) {
-            AndroidSessionWrapper asw = AndroidSessionWrapper.mockEasiestRoute(platform, r.getFormNamespace(), caseIDs[0]);
-            asw.setFormRecordId(parsed.getID());
-            
-            SqlStorage<SessionStateDescriptor> ssdStorage = CommCareApplication._().getUserStorage(SessionStateDescriptor.class);
-            
-            //Also bad: this is not synchronous with the parsed record write
-            try {
-                ssdStorage.write(asw.getSessionStateDescriptor());
-            } catch (StorageFullException e) {
-            }
-        }

```

Comment:
Move SessionStateDescriptor storage write out of this method and into its caller, who also saves the FormRecord to storage. These writes should happen as a transaction, so this makes it easier to see that they are linked